sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +411,237 @@ class GpuSequenceMeta(
+        }
+        withResource(allDown) { _ =>
+          require(isAllValidTrue(allDown),
+              "Illegal sequence boundaries: step < 0 but start < stop")
+        }
+
+        // (step == 0 && start == stop)
+        val equalTbl = withResource(step.equalTo(zero)) { zeroStep =>
+          startStopTable.filter(zeroStep)
+        }
+        val allEq = withResource(equalTbl) { _ =>
+          equalTbl.getColumn(0).equalTo(equalTbl.getColumn(1))
+        }
+        withResource(allEq) { _ =>
+          require(isAllValidTrue(allEq),
+              "Illegal sequence boundaries: step == 0 but start != stop")
+        }
+      }
+    } // end of zero
+
+    // Spark's algorithm to get the length (aka size)
+    // ``` Scala
+    //   size = if (start == stop) 1L else 1L + (stop.toLong - start.toLong) / estimatedStep.toLong
+    //   require(size <= MAX_ROUNDED_ARRAY_LENGTH,
+    //       s"Too long sequence: $len. Should be <= $MAX_ROUNDED_ARRAY_LENGTH")
+    //   size.toInt
+    // ```
+    val sizeAsLong = withResource(Scalar.fromLong(1L)) { one =>
+      val diff = withResource(stop.castTo(DType.INT64)) { stopAsLong =>
+        withResource(start.castTo(DType.INT64)) { startAsLong =>
+          stopAsLong.sub(startAsLong)
+        }
+      }
+      val quotient = withResource(diff) { _ =>
+        withResource(step.castTo(DType.INT64)) { stepAsLong =>
+          diff.div(stepAsLong)
         }
-        val size = closeOnExcept(step) { _ =>
-          withResource(difference.abs()) { absDifference =>
-            absDifference.add(one)
+      }
+      // actualSize = 1L + (stop.toLong - start.toLong) / estimatedStep.toLong
+      val actualSize = withResource(quotient) { quotient =>
+        quotient.add(one, DType.INT64)
+      }
+      withResource(actualSize) { _ =>
+        val mergedEquals = withResource(start.equalTo(stop)) { equals =>
+          if (step.hasNulls) {
+            // Also set the row to null where step is null.
+            equals.mergeAndSetValidity(BinaryOp.BITWISE_AND, equals, step)
+          } else {
+            equals.incRefCount()
           }
         }
-        Seq(size, step)
+        withResource(mergedEquals) { _ =>
+          mergedEquals.ifElse(one, actualSize)
+        }
       }
     }
-  }
-
-  override def doColumnar(start: GpuColumnVector, stop: GpuColumnVector): ColumnVector = {
-    withResource(calculateSizeAndStep(start.getBase, stop.getBase, start.dataType())) { ret =>
-      ColumnVector.sequence(start.getBase, ret(0), ret(1))
-    }
-  }
 
-  override def doColumnar(start: GpuScalar, stop: GpuColumnVector): ColumnVector = {
-    withResource(calculateSizeAndStep(start.getBase, stop.getBase, stop.dataType())) { ret =>
-      withResource(ColumnVector.fromScalar(start.getBase, stop.getRowCount.toInt)) { startV =>
-        ColumnVector.sequence(startV, ret(0), ret(1))
+    // check size
+    closeOnExcept(sizeAsLong) { _ =>
```

Comment:
Rather than have a `closeOnExcept(sizeAsLong)` immediately followed by a `withResource(sizeAsLong)`, we can just combine the two blocks together in a single `withResource(sizeAsLong)`.  Is there any advantage to separate the `closeOnExcept` block?

Reply:
No, just thought it was clearer to read. Updated